packages/neo-one-client/src/user/LocalUserAccountProvider.ts

```diff
@@ -823,6 +841,172 @@ export class LocalUserAccountProvider<TKeyStore extends KeyStore, TProvider exte
+    readonly from: UserAccountID;
+    readonly monitor?: Monitor;
+    readonly byteLimit: number;
+  }): Promise<TransactionModel> {
+    const newInputSize = 68;
+    const newOutputSize = 120;
+    const messageSize = transactionUnsignedIn.serializeUnsigned().toString('hex').length;
+
+    if (byteLimit - messageSize < 0) {
+      throw new TransactionSizeExceedsLimitError(messageSize, byteLimit);
+    }
+    if (byteLimit - messageSize - newInputSize < 0) {
+      return transactionUnsignedIn; // if we can't even add 1 input don't bother...
+    }
+
+    let updatedOutputs = transactionUnsignedIn.outputs;
+    let newOutputs: ReadonlyArray<OutputModel> = [];
+    let newInputs: ReadonlyArray<InputModel> = [];
+    let looseAssets: { [key: string]: ReadonlyArray<UnspentOutput> } = {};
+
+    const createConsolidatedTransaction = (): TransactionModel => {
+      switch (transactionUnsignedIn.type) {
+        case TransactionType.Contract:
+          return new ContractTransaction({
+            version: transactionUnsignedIn.version,
+            attributes: transactionUnsignedIn.attributes,
+            inputs: transactionUnsignedIn.inputs.concat(newInputs),
+            outputs: updatedOutputs.concat(newOutputs),
+            scripts: transactionUnsignedIn.scripts,
+            hash: transactionUnsignedIn.hash,
+          });
+        case TransactionType.Claim:
+          return new ClaimTransaction({
+            version: transactionUnsignedIn.version,
+            attributes: transactionUnsignedIn.attributes,
+            inputs: transactionUnsignedIn.inputs.concat(newInputs),
+            outputs: updatedOutputs.concat(newOutputs),
+            scripts: transactionUnsignedIn.scripts,
+            hash: transactionUnsignedIn.hash,
+            claims: transactionUnsignedIn.claims,
+          });
+        case TransactionType.Issue:
+          return new IssueTransaction({
+            version: transactionUnsignedIn.version,
+            attributes: transactionUnsignedIn.attributes,
+            inputs: transactionUnsignedIn.inputs.concat(newInputs),
+            outputs: updatedOutputs.concat(newOutputs),
+            scripts: transactionUnsignedIn.scripts,
+            hash: transactionUnsignedIn.hash,
+          });
+        default:
+          return transactionUnsignedIn;
+      }
+    };
+
+    const getMessageSize = () => messageSize + newInputs.length * newInputSize + newOutputs.length * newOutputSize;
+    const updateOutput = (output: OutputModel, amount: BigNumber): OutputModel =>
+      new OutputModel({
+        asset: output.asset,
+        value: output.value.add(clientUtils.bigNumberToBN(amount, 8)),
+        address: output.address,
+      });
+
+    const consoladatableUnspentsIn = await this.provider.getUnspentOutputs(from.network, from.address, monitor);
+    consoladatableUnspentsIn
+      .filter(
+        (unspent) =>
+          !inputs.some((input) => unspent.txid === input.txid && unspent.vout === input.vout) &&
```

Comment:
This has complexity of O(n^2). Stick them in a set similar to `mutableUsedOutputs` and filter that way.